Migrate from Slack to Discord

jbboynton opened the pull request:
Update the project so that notifications and messages are posted to
their new channel in the Land of Fires Discord server.

Discord has a very similar structure to Slack—both services allow
messages to be created via webhooks. So, the bulk of this migration
comes down to updating webhook URLs and renaming "Slack" to "Discord".

This update includes a migration, which prompts the user to input the
new Discord endpoints on the command line.

jbboynton committed:
Duplicate and rename Slack.py to Discord.py

jbboynton committed:
Find/replace Slack with Discord

jbboynton committed:
Add migration for v1.4

jbboynton committed:
Remove Slack.py

jbboynton committed:
Migrate from Slack to Discord

Update the project so that notifications and messages are posted to
their new channel in the Land of Fires Discord server.

Discord has a very similar structure to Slack—both services allow
messages to be created via webhooks. So, the bulk of this migration
comes down to updating webhook URLs and renaming "Slack" to "Discord".

This update includes a migration, which prompts the user to input the
new Discord endpoints on the command line.

jbboynton committed:
Add missing import in migration

jbboynton committed:
Rewrite Discord.py

This class was created by duplicating the old `Slack.py` class without
modification. When I began to use the new `Discord` class, I started to
realize that the old `Slack` class had some bloat to it. For example,
I noticed there was a some manual HTTP handling and hand-rolled retry
behavior. I also noticed that much of the class was checking
`Dev.isDev()`, and there were a lot of places where the environment was
being checked.

I decided that it would be a good opportunity to refresh the class and
try to strip it down to the bare essentials. For example, I moved a lot
of data into attributes, such as the environment state. This allowed me
to simplify a lot of the environment checking. I also changed the
constructor behavior so that instantiating a new object wouldn't have
the side effect of firing the HTTP request—firing the request was moved
to an instance method.

Speaking of HTTP—I decided to completely punt on that. I found a very
simple `discord-webhooks` library and using that seemed like a much,
much better idea than learning Discord's webhook APIs.

There is still room for improvement, and there's also the chance that
some of what I removed will need to go back in. However, I think this is
a solid step in the right direction, and sets a good precedent for class
design going forward.

jbboynton committed:
Refactor Discord.py

I didn't feel great about the half-refactoring I started in the last
commit, so I decided to keep working with it until I felt it was in a
better place.

Now, the `Discord` class has a much clearer, singular focus. I was able
to push the other concerns out of this class, such as file I/O, Drive
URL string construction, and settings initialization. These things have
been moved to places that I think make sense (with the exception of the
log file I/O, which really needs to be wrapped in its own class, but
that's a different issue entirely).

The `Discord` class now has a predictable and simple API. `Discord`
instances have three main methods which can be used by the caller to
post stuff in the Discord channel. The methods are `post_message`,
`post_link`, and `post_log`. Each behave slightly differently and have
minor parameter differences. This pattern will be easy to continue
developing as the needs of the app change.

jbboynton commented on migrations/1_4.py:
```diff
@@ -0,0 +1,66 @@
+    if not Settings.get_key("update_git"):
+        Settings.set_key("update_git",True)
+
+        dialog = Dialog(
+            title = "Upgrading to V1.4!",
+            body  = [
+                f'A simple, innocent update that will switch SAM from Slack to Discord.',
+                f'\n', f'\n',
+                f'This update requires a manual change to the .settings file ',
+                f'in this project. The old Slack endpoint keys and values ',
+                f'need to be replaced with the new Discord endpoint keys and ',
+                f'values. The new keys and values can be found in this file: ',
+                f'\n', f'\n',
+                f'   https://drive.google.com/file/d/1Q8KppHtX1ZW4ZJUIDa6fhQ-YxE44Hxkn/view?usp=sharing',
+                f'\n', f'\n',
+                f'If you run into difficulty with this manual update, post a ',
+                f'message in Discord and someone can assist.',
+                f'\n', f'\n',
+                f'For more information, please refer to the repository commits and/or the Discord channel.',
+                f'\n', f'\n',
+            ]
+        )
+        dialog.press_enter()
```
@vindennl48 It seems to me that the migrations may be irreversible and unrunnable. I believe there's at least one migration that calls a method that was removed a while ago, which ultimately prevents new migrations from running.

Other than a `git pull`, this migration doesn't require a lot of changes to existing files. The only change that should ideally be done with a migration is adding the new webhook URLs to the `.settings` file. But since new migrations can't be run, I think manual intervention is really the only way to go here.

vindennl48 replied:
You are correct that this migration system is one direction. The idea behind this was, when someone goes to install this, say right now, they would be completely up-to-date. The migrations only run if your version number is less than a migration file. So if this was installed now, they should never run.

The only time it can become an issue is if someone installs it now, then doesn't run it for 2 upgrades. That could be an issue. That would need some updating. However, since I communicate with y'all on the reg. about that stuff, I try and force everyone to upgrade before I push out any new migrations.

So with that being said, this migration should work fine.

jbboynton committed:
Consolidate log file I/O in Log.py

jbboynton commented on migrations/1_4.py:
```diff
@@ -0,0 +1,66 @@
+    if not Settings.get_key("update_git"):
+        Settings.set_key("update_git",True)
+
+        dialog = Dialog(
+            title = "Upgrading to V1.4!",
+            body  = [
+                f'A simple, innocent update that will switch SAM from Slack to Discord.',
+                f'\n', f'\n',
+                f'This update requires a manual change to the .settings file ',
+                f'in this project. The old Slack endpoint keys and values ',
+                f'need to be replaced with the new Discord endpoint keys and ',
+                f'values. The new keys and values can be found in this file: ',
+                f'\n', f'\n',
+                f'   https://drive.google.com/file/d/1Q8KppHtX1ZW4ZJUIDa6fhQ-YxE44Hxkn/view?usp=sharing',
+                f'\n', f'\n',
+                f'If you run into difficulty with this manual update, post a ',
+                f'message in Discord and someone can assist.',
+                f'\n', f'\n',
+                f'For more information, please refer to the repository commits and/or the Discord channel.',
+                f'\n', f'\n',
+            ]
+        )
+        dialog.press_enter()
```
@vindennl48 It seems to me that the migrations may be irreversible and unrunnable. I believe there's at least one migration that calls a method that was removed a while ago, which ultimately prevents new migrations from running.

Other than a `git pull`, this migration doesn't require a lot of changes to existing files. The only change that should ideally be done with a migration is adding the new webhook URLs to the `.settings` file. But since new migrations can't be run, I think manual intervention is really the only way to go here.

vindennl48 replied:
You are correct that this migration system is one direction.  The idea behind this was, when someone goes to install this, say right now, they would be completely up-to-date.  The migrations only run if your version number is less than a migration file.  So if this was installed now, they should never run.  

The only time it can become an issue is if someone installs it now, then doesn't run it for 2 upgrades.  That could be an issue.  That would need some updating.  However, since I communicate with y'all on the reg. about that stuff, I try and force everyone to upgrade before I push out any new migrations.

So with that being said, this migration should work fine.

jbboynton replied:
Word. The situation I've run into at least 2 or 3 times is that I need to reinstall the entire project from scratch. Definitely something we can figure out though.

vindennl48 replied:
hmm yeah that shouldn't be happening and during my tests hasn't been something I've experienced.  Not sure what is different but we definitely should dig deeper into that then at some point.

jbboynton commented on .main.py:
```diff
@@ -1,5 +1,5 @@
 import traceback,sys
-from src.Slack import Slack
+from src.Discord import Discord
```
@vindennl48 Is this import needed? Doesn't look like the class is being used here at all.

jbboynton replied:
I went ahead and pulled this import out—pretty sure it's not needed.

vindennl48 replied:
I agree, these can be removed.  I think they used to..

jbboynton commented on src/Discord.py:
```diff
@@ -0,0 +1,69 @@
+from src.Settings import Settings
+from src.TERMGUI.Log import Log
+
+from discord_webhook import DiscordWebhook
```
@vindennl48 I took my best guess at formatting and style but I'm way out of my element here. Please feel free to make suggestions or tweaks.

jbboynton commented on src/Settings.py:
```diff
@@ -74,3 +74,38 @@ def get_username(capitalize=False):
 
         return username
+
+    def check_discord_endpoints():
+        settings = Settings.get_all()
+
+        if not Settings.discord_prod_key in settings or \
+           not Settings.discord_dev_key in settings:
+            return False
+
+        if settings[Settings.discord_prod_key] == "" or \
+           settings[Settings.discord_dev_key] == "":
+            return False
+
+        return True
+
+    def set_discord_endpoints():
+        if not Settings.check_discord_endpoints():
+            drive = Drive()
+
+            Settings.set_key(
+                key  = Settings.discord_prod_key,
+                data = drive.get_json_key(
+                    remote_file    = f'{LOFSM_DIR_PATH}/db.json',
+                    local_filepath = f'temp/db.json',
+                    key            = Settings.discord_prod_key
+                )
+            )
+
+            Settings.set_key(
+                key  = Settings.discord_dev_key,
+                data = drive.get_json_key(
+                    remote_file    = f'{LOFSM_DIR_PATH}/db.json',
+                    local_filepath = f'temp/db.json',
+                    key            = Settings.discord_dev_key
+                )
+            )
```
@vindennl48 I couldn't really find a better place for all of this, but it seems like it makes more sense here than in the `Discord` class, I think? I'm thinking that this code really just runs once when the program initializes to set some state—maybe we could extract an `Init.py` and have it contain all the setup code?

vindennl48 replied:
I think since it falls under the Discord umbrella, it should live in the Discord class. Similar to how it was with the Slack class.

jbboynton replied:
I would disagree. In my opinion, this group of methods does not fall under the Discord umbrella. It falls under the umbrella of setting and verifying variables read from the environment.

However, I would agree that `Settings.py` is not necessarily the right place for it. I would like to see all of this code moved into a `Config.py`, `Env.py`, or similar module. This module should run once at the start of the program. It should set any variables read from `.settings` and verify that all required values are present up front.

I'm happy to start working on this refactoring, but I think the right time to do it would be after this PR has been merged. Would you agree?

vindennl48 replied:
Not trying to be difficult, but I do not agree.  I think you are bringing a Rails perspective to this when it doesn't necessarily fit (imo).  The last time I refactored this project, I consciously made the decision to have all things that pertain to itself in it's own namespace/class.  I actually had pieces like this spread all over and it became a problem.  Reason being, I don't touch this more than maybe once a year / once every 2 years.  It's much easier to come in and look in the 'Discord' class file and have everything pertaining to discord there instead of having to remember where all the little chunks of discord lay within the project.  

It's setup similar to how the C++ programs work.  Every device/section/topic/(however you want to classify something) is grouped in namespaces or classes.  This way it is easier to find the code in one spot.

Not saying one way is better than the other, just that I haven't used rails in a long time, I'm much more suited to this style and I think it makes more sense specifically for this project because of how little we need to maintain it.

jbboynton committed:
Bug fixes

vindennl48 commented on src/Discord.py:
```diff
@@ -0,0 +1,69 @@
+    dev_endpoint = Settings.get_key(Settings.discord_dev_key)
+    prod_endpoint = Settings.get_key(Settings.discord_prod_key)
+
+    def __init__(self, content = "", quiet = False):
```
Since you changed the way we interact with this class, I don't think we need the content argument here anymore?

jbboynton replied:
No, it's not necessary. I was kind of thinking it would be better if we were passing all the text to the constructor but it's not practical. I'll remove it for now.

jbboynton committed:
Remove unnecessary imports

vindennl48 commented on src/Discord.py:
```diff
@@ -0,0 +1,69 @@
+            Log("Notification webhook failed:")
+            Log(f"  Webhook: {inspect.getmembers(webhook)}")
+            Log(f"  Response: {inspect.getmembers(response)}")
+
```
Can you add a line to the log saying that the notification was successful as well as the message text.

vindennl48 commented on src/Discord.py:
```diff
@@ -0,0 +1,69 @@
+        response = webhook.execute()
+
+        if not self.quiet and not response.ok:
+            Log("Notification webhook failed:")
```
I have some options for the Log().  second argument gives you different leader types.  for line 40, you can use 'warning' and lines 40,41 you can use 'sub'

jbboynton committed:
PR comment: remove unneeded parameter

vindennl48 commented on src/Discord.py:
```diff
@@ -0,0 +1,69 @@
+    dev_endpoint = Settings.get_key(Settings.discord_dev_key)
+    prod_endpoint = Settings.get_key(Settings.discord_prod_key)
+
+    def __init__(self, content = "", quiet = False):
```
Since you changed the way we interact with this class, I don't think we need the content argument here anymore?

jbboynton replied:
[Already removed](https://github.com/vindennl48/LOFSongManager/pull/10/commits/834d8786b6328d472e5c4bfa0b665cfb602a443c) 👍

vindennl48 commented on src/Discord.py:
```diff
@@ -0,0 +1,69 @@
+        self.post()
+
+    def post_log(self, content, quiet = True):
+        log = f"{self.username} – Log: {content}"
```
Not a big one but if we can separate the "<username> - Log:" line and start the content on the next line?  So that the log file's first line isn't indented with the username.

jbboynton replied:
Would this suffice?

```suggestion
        log = f"Log – {self.username}:\n\n{content}"
```

My thought was that it would be good to have the username and log content all in the message, so that it's never unclear whose log you're reading in Discord, and so that we can avoid making two HTTP requests. But this is definitely a micro optimization, and I have no strong feelings either way. Let me know if you prefer it how you had it originally with the two requests back-to-back.

jbboynton replied:
So I went ahead and made [this change](https://github.com/vindennl48/LOFSongManager/pull/10/commits/76f191c1ff90b83e61ad178090642f09603546ab) just to get it in there tonight, but still let me know if you'd prefer something different.

vindennl48 commented on src/Project/Compress.py:
```diff
@@ -91,10 +91,11 @@ def move_extracted_song_to_temp(self):
+                link_title = f'{audio_type} for {self.entry.name}, "{mp3.name}"'
+                url        = Drive.file_url(mp3_id)
+
+                Discord().post_link(link_title, url)
```
I'd rather have it all encompassed in the parentheses like I had the Slack function but if you want it like this I won't push back too badly.

jbboynton replied:
This one I do have strong feelings about. The thing I really want to avoid is having implicit behavior in the constructor. In my opinion, the constructor shouldn't really _do_ anything other than return an instance of the class. My understanding of good object-orientated design is that an object's behavior should be defined in its methods. In my experience, this tends to work out well over the long-term, and the benefits outweigh the drawbacks (namely, more verbose and ugly code). I can definitely elaborate on some of the specific benefits if you're not fully sold on this.

All of that said, I think we can still improve this. Take a look at [this commit](https://github.com/vindennl48/LOFSongManager/pull/10/commits/cd8798947b47eb06f52d2d5ce1fbae08b2c6270a). Instead of having to instantiate a new object with `Discord()` and then chaining a `post_*` method every time, I thought it would be cleaner to have a single class method that would accept a dict of options and data. The result is that notifications are created by calling `Discord.notify({ ... })`.

It's still not perfect, but I think it makes for a better interface, and it feels more natural than how it was before. What do you think?

vindennl48 replied:
I think we are discussing two different topics.  My comment here was about these two options:
```
link_title = f'{audio_type} for {self.entry.name}, "{mp3.name}"'
url        = Drive.file_url(mp3_id)

Discord().post_link(link_title, url)
```

as opposed to:

```
Discord().post_link(
    f'{audio_type} for {self.entry.name}, "{mp3.name}"',
    Drive.file_url(mp3_id)
)
```

The topic you are referring to I mentioned in another comment.  I will address it there.

vindennl48 commented on src/Project/Upload.py:
```diff
@@ -55,7 +55,9 @@ def upload_project(self):
-            Slack(f'{Slack.get_nice_username()} uploaded a new version of {Slack.make_nice_project_name(self.entry.name)}')
+            content = f"{Discord.get_nice_username()} uploaded a new version " \
+                    "of {Discord.make_nice_project_name(self.entry.name)}"
+            Discord().post_message(content)
```
Same thing here as in the Compress.py file.

vindennl48 commented on src/Update.py:
```diff
@@ -53,7 +54,9 @@ def run_migrations():
+                # Push a notification to Discord
+                content = f"{Settings.get_username(capitalize=True)} has " \
+                        "upgraded to V{migration_version}!"
+                Discord().post_message(content)
```
Same comment here as in Compress.py

vindennl48 commented on src/Update.py:
```diff
@@ -41,7 +41,8 @@ def run_migrations():
 
                 if ans == 0:
-                    Slack.upload_log()
+                    Discord().post_log(Log.dump())
```
Do we necessarily need the parentheses next to Discord() <= ? I wrote the Slack class static so they did not have to be used.  Again not a huge deal but I believe I have most of this program written with that in mind.

jbboynton replied:
Does [the addition of the `Discord.notify()` class method](https://github.com/vindennl48/LOFSongManager/pull/10/commits/cd8798947b47eb06f52d2d5ce1fbae08b2c6270a) resolve this?

vindennl48 replied:
I pulled your comment from the other comment since it essentially is about this.
> James -
> This one I do have strong feelings about. The thing I really want to avoid is having implicit behavior in the constructor. In my opinion, the constructor shouldn't really do anything other than return an instance of the class. My understanding of good object-orientated design is that an object's behavior should be defined in its methods. In my experience, this tends to work out well over the long-term, and the benefits outweigh the drawbacks (namely, more verbose and ugly code). I can definitely elaborate on some of the specific benefits if you're not fully sold on this.

> All of that said, I think we can still improve this. Take a look at [this commit](https://github.com/vindennl48/LOFSongManager/pull/10/commits/cd8798947b47eb06f52d2d5ce1fbae08b2c6270a). Instead of having to instantiate a new object with Discord() and then chaining a post_* method every time, I thought it would be cleaner to have a single class method that would accept a dict of options and data. The result is that notifications are created by calling Discord.notify({ ... }).

> It's still not perfect, but I think it makes for a better interface, and it feels more natural than how it was before. What do you think?

I believe the disconnect here is that I'm using the class as a static class, basically imitating how C++ works with Namespaces.  Because they do not need an instance, they should be global throughout the project or else we will get all sorts of circular dependency issues.  The other reason I had done it this way (I did the same for the Log class as well) was to make things easier to read and remember how things worked.  Instead of having to memorize the member functions to make something happen, it is built inherently into the 'constructor' (loosely used term here because of how I am using the class).  It was a conscious decision to build it that way at the time.

I'm not too keen on refactoring the whole project again when it isn't necessarily needed.  I would like everything the same so if this class changes, then they all should be changed.

vindennl48 commented on src/Project/Upload.py:
```diff
@@ -55,7 +55,9 @@ def upload_project(self):
             self.entry.update()
 
-            Slack(f'{Slack.get_nice_username()} uploaded a new version of {Slack.make_nice_project_name(self.entry.name)}')
+            content = f"{Discord.get_nice_username()} uploaded a new version " \
```
I don't believe 'Discord.get_nice_username' exists?

jbboynton replied:
You are correct—this was a find/replace bug that is now fixed.

vindennl48 commented on src/menus/menu_create_project.py:
```diff
@@ -34,7 +35,7 @@ def menu_create_project():
             Folder.create( new_song.parent/location )
 
-        Log(f'New song "{Slack.make_nice_project_name(name)}" created!', "notice")
+        Log(f'New song "{nice_name}" created!', "notice")
```
I think because this is used in multiple places, this should live somewhere as a function to make the nice project name.  I'm OK with sticking it in the Project class it's self or in the Discord class.

vindennl48 replied:
I just noticed it also exists in Upload.py

jbboynton replied:
That was my first instinct as well. Unfortunately, as soon as I tried to hang new methods off the `Project` class, I started getting a bunch of import errors due to circular dependencies. I started to try to fix the errors, but pretty quickly I realized that this is kind of a big issue.

So, for now, there is some duplication. I think the best thing to do is live with it for now, don't let it get any worse, and refactor module importing in the near future. Will that work for you?

vindennl48 commented on src/Project/Upload.py:
```diff
@@ -55,7 +55,9 @@ def upload_project(self):
 
-            Slack(f'{Slack.get_nice_username()} uploaded a new version of {Slack.make_nice_project_name(self.entry.name)}')
+            content = f"{Discord.get_nice_username()} uploaded a new version " \
+                    "of {Discord.make_nice_project_name(self.entry.name)}"
```
Actually same here with make_nice_project_name.  I made a comment about this in 'menu_create_project.py'

vindennl48 reviewed (changes requested):
Looks good, just a few minor things.

jbboynton committed:
PR comment: log successful webhook requests

[Link to comment](https://github.com/vindennl48/LOFSongManager/pull/10#discussion_r928167529)

jbboynton committed:
PR comment: use consistent log formatting

[Link to comment](https://github.com/vindennl48/LOFSongManager/pull/10#discussion_r928167640)

jbboynton committed:
PR comment: add `Discord.notify` class method

[Link to comment](https://github.com/vindennl48/LOFSongManager/pull/10#discussion_r928167960)

Create a single class method to replace the `post_message`, `post_link`,
and `post_log` instance methods.

jbboynton committed:
Bug fix: eliminate circular dependency I created

jbboynton committed:
Formatting fix: use consistent style and undo unnecessary churn

jbboynton committed:
PR comment: fix log formatting

[Link to comment](https://github.com/vindennl48/LOFSongManager/pull/10#discussion_r928167762)

jbboynton commented:
Replying to [your comment in `Settings.py`](https://github.com/vindennl48/LOFSongManager/pull/10#discussion_r9281681260):

>I think since it falls under the Discord umbrella, it should live in the Discord class. Similar to how it was with the Slack class.

I would disagree. In my opinion, this group of methods does not fall under the Discord umbrella. It falls under the umbrella of setting and verifying variables read from the environment.

However, I would agree that `Settings.py` is not necessarily the right place for it. I would like to see all of this code moved into a `Config.py`, `Env.py`, or similar module. This module should run once at the start of the program. It should set any variables read from `.settings` and verify that all required values are present up front.

I'm happy to start working on this refactoring, but I think the right time to do it would be after this PR has been merged. Would you agree?

jbboynton commented:
After some discussion with @vindennl48 we both feel that this PR too disruptive as currently written. Closing this and opening [this](https://github.com/vindennl48/LOFSongManager/pull/11).
